btree/btree.c

```diff
@@ -3688,28 +3696,28 @@ static int64_t btree_get_tick(struct m0_sm_op *smop)
 			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
 		}
 	case P_LOCK:
-		if (!locked(tree))
+		if (!oi->i_lock_aquired)
 			return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
-					    bop->bo_arbor->t_desc, P_CHECK);
+					    bop->bo_arbor->t_desc, oi, P_CHECK);
 		else
 			return P_CHECK;
 	case P_CHECK:
-		oi->i_trial++;
 		if (!path_check(oi, tree, &bop->bo_rec.r_key.k_cookie)) {
 			if (oi->i_trial == MAX_TRIALS) {
 				if (bop->bo_flags & BOF_LOCKALL)
 					return fail(bop, -ETOOMANYREFS);
 				else
 					bop->bo_flags |= BOF_LOCKALL;
+			} else {
+				oi->i_trial++;
+				lock_op_unlock(tree, oi);
 			}
 			if (bop->bo_arbor->t_height != tree->t_height) {
-				/* If height increased */
-				lock_op_unlock(tree);
+				/* If height has changed. */
 				return m0_sm_op_sub(&bop->bo_op, P_CLEANUP,
 				                    P_INIT);
```

Comment:
change `P_INIT to P_SETUP`

Reply:
then we won't be able to check cookies while traversing again

Reply:
done